src/recovery/store/in_mem.rs

```diff
@@ -331,58 +331,201 @@ fn write_discard_drops_key() {
 
 /// A progress store with all data in-memory.
 #[derive(Debug)]
-pub(crate) struct InMemProgress(HashMap<WorkerKey, BorderEpoch>);
+pub(crate) struct InMemProgress {
+    /// Stores the current execution.
+    ///
+    /// Defaults to `-1` so the initial call to [`resume_from`] will
+    /// generate the execution `0`. There will never be any recovery
+    /// progress data that will match execution `-1`, so as soon as we
+    /// start reading any, we'll bump to the correct execution.
+    ex: i128,
+    frontiers: HashMap<WorkerIndex, WorkerFrontier>,
+}
 
 impl InMemProgress {
-    pub(crate) fn new() -> Self {
-        Self(HashMap::new())
+    /// Init to the "default" progress of a cluster of a given
+    /// size.
+    ///
+    /// You have to specify a worker count so the initial data is
+    /// well-formed.
+    pub(crate) fn new(count: WorkerCount) -> Self {
+        let ex = Into::<i128>::into(<u64 as Timestamp>::minimum()) - 1;
+        let mut frontiers = HashMap::new();
+        for worker in count.iter() {
+            frontiers.insert(worker, WorkerFrontier(<u64 as Timestamp>::minimum()));
+        }
+        Self { ex, frontiers }
     }
-}
 
-impl InMemProgress {
-    /// Calculate the resume epoch from the previous cluster state.
+    pub(crate) fn worker_count(&self) -> WorkerCount {
+        WorkerCount(self.frontiers.len())
+    }
+
+    fn init(&mut self, ex: Execution, count: WorkerCount, epoch: ResumeEpoch) {
+        // Tidbit: you almost never want to use `as` numerical
+        // casts. They do "bit folding" on downcasts and don't panic
+        // when you're out of range, which is probably what you want.
+        let in_ex: i128 = ex.0.into();
+        // Execution should never regress because the entire cluster
+        // is shut down before resuming.
+        assert!(!(in_ex < self.ex), "Execution regressed");
+        if in_ex > self.ex {
+            self.ex = in_ex;
+            self.frontiers.clear();
+            for worker in count.iter() {
+                self.frontiers.insert(worker, WorkerFrontier(epoch.0));
+            }
```

Comment:
Similar to the above, you could write:
```rust
            self.frontiers = count
                .iter()
                .map(|worker| (worker, WorkerFrontier(epoch.0)))
                .collect();

```

Not sure it doing this way implies an additional allocation though

Reply:
Not in a hot loop anyway, so easier to read.